aten/src/ATen/native/cuda/Indexing.cu

```diff
@@ -194,34 +204,38 @@ void index_put_accum_kernel(Tensor & self, const c10::List<c10::optional<Tensor>
   std::vector<int64_t> inversePerm;
   std::tie(linearIndex, src, nElemBefore, strideBefore, sliceSize, inversePerm) = makeLinearIndex(self, indices, !unsafe);
   int64_t num_indices = linearIndex.numel();
+
+  TORCH_CHECK(num_indices <= std::numeric_limits<int>::max(),
+    "index_put of tensors larger than INT_MAX is not supported yet in pytorch");
+
   if (num_indices > 0 && sliceSize > 0) {
       const bool permuted = !src.is_contiguous();
       auto src_ = permuted ? src.contiguous() : src;
       linearIndex = linearIndex.reshape(-1);
       auto sorted_indices = at::empty_like(linearIndex, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
       auto orig_indices = at::empty_like(linearIndex, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
-      using device_ptr = thrust::device_ptr<int64_t>;
       const cudaStream_t stream = at::cuda::getCurrentCUDAStream();
 
       linearIndex.divide_(sliceSize, "trunc");
+
+      // cub on CUDA <= 11.2 have a bug that for small sizes
+      // cub's sort can be much slower than thrust's merge sort
+      // this bug is fixed in CUDA 11.3
+#if defined(CUDA_VERSION) && CUDA_VERSION < 11030
+      if (num_indices < 50000) {
+        index_put_accum_kernel_thrust_helper(linearIndex, orig_indices, sorted_indices, num_indices);
+      } else
+#endif
       {
-      sorted_indices.copy_(linearIndex);
-      auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
-      auto policy = thrust::cuda::par(allocator).on(stream);
-
-      // Fill sortedOrigIndices with sequential indices
-      const auto count_iter = thrust::counting_iterator<int64_t>(0);
-      auto orig_data = device_ptr(orig_indices.data_ptr<int64_t>());
-      thrust::copy(policy, count_iter, count_iter + num_indices, orig_data);
-
-      // Sort the inputs into sorted with the corresponding indices; we
-      // don't need a stable or multidimensional sort, so just use Thrust
-      // directly
-      // Sort; a stable sort is not required
-      // NB - not passing comparator causes thrust to use radix sort, and it hurts perf A LOT, at least for medium (few K) sized indices
-      auto sorted_data = device_ptr(sorted_indices.data_ptr<int64_t>());
-      thrust::sort_by_key(policy, sorted_data, sorted_data + num_indices, orig_data, ThrustLTOp<int64_t>());
+      // Sort the inputs into sorted with the corresponding indices
+      auto range = at::arange(num_indices, linearIndex.options());
+      int64_t nbits = std::max<int64_t>(0, std::ceil(std::log2(double(largestIndex(self)) / sliceSize))) + 1;
```

Comment:
this computation seems off. Cub sorts signed by xoring the highest bit https://github.com/NVIDIA/cub/blob/a8910accebe74ce043a13026f8e71d678cddd6c1/cub/util_type.cuh#L1017, so technically as you said one cannot set nbits for sorting signed. Also as we’ve discussed, it’s not a real problem for index_put because index_put does not accept negative indices (does it not? I can never remember which functions wrap indices, and which don’t), but it also means that the bit computation is off, int_max should sort on 31 bits, and 0-15 should sort on 4 bits, whereas formula gives 32 and 5 respectively.